### PR TITLE
Game server api

### LINKED ISSUE #19853: Game server admin API
An API that makes certain server specific information and actions available. Needs to require some form of authentication.

This would allow for a way for things like banning someone on one server to kick them from other servers, or banning someone from SS14.Admin to kick them from servers. Also would allow a dashboard to be created to let admins see things like which servers currently have no admins.

# Available information

- [ ] Round number
- [ ] Connected players
- [ ] Active admins
- [ ] Active game rules
- [ ] Active game preset
- [ ] Active map
- [ ] MOTD
- [ ] Panic bunker status

# Available actions

- [ ] End/restart round
- [ ] Kick player
- [ ] Add/remove game rule
- [ ] Set game preset
- [ ] Set map
- [ ] Set MOTD
- [ ] Set panic bunker status

# Explicitly unavailable actions/information
This is stuff that probably shouldn't be done through the API since it can be done some other way, so it should probably not be available through the API

- Anything you can get from the database
  - Bans
  - Notes
  - Playtime
  - Logs
  - Permissions
- Anything you can do from the database
  - Ban
  - Note

### PR DESCRIPTION
**Pull request by Simyon264**
## About the PR
Adds an admin API which allows for better moderation.
## Why / Balance

resolves #19853 

## Todo
- [x] Some of the code is very messy. I need to un-messify it
- [x] Video showcasing the API

## Technical details
I define the `ServerApi` class and register it in the IoCManager in `ServerContentIoC`. This class is used to handle HTTP requests to the /admin endpoint.
The class has 2 helper methods. `RunOnMainThread` and `CheckAccess`.
`RunOnMainThread` is a helper function for executing a task on the main thread asynchronously.
`CheckAccess`  checks the authorization header to determine if the requester has the correct token for accessing the admin API.
Methods prefixed with `Action` are, well, actions that can be performed through the admin API. These include actions for changing Panic Bunker settings, setting the MOTD, forcing presets, ending and adding game rules, and kicking players.
There are three Methods which handle fetching Information. These don't have a fancy prefix.
`GetForcePresets` returns all instances of `GamePresetPrototype`.
`GetGameRules` returns all `EntityPrototype` with the `GameRuleComponent`.
`InfoHandler` returns general information about the current server status. Look at the Media section for more information.

Everything touching in-game logic, should be ran on the main thread. I have probably missed something, so please mentioned those I have missed, so that I can make them not explode the server.
## Media
- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase
https://cdn.discordapp.com/attachments/903694002577113118/1190028451533553825/Desktop_2023.12.28_-_21.06.37.04.mp4?ex=65a04f5c&is=658dda5c&hm=e9a0cbe6831c41fee58ae9e9914ed2ff9a8d06542a07a39a805b37fb972ede2d&

(I don't know how to embed video)
## Breaking changes
None

### CODE REVIEW

**Comment by VasilisThePikachu** [comment]
Btw round ID is already being sent in /status its not needed
image

**Comment by VasilisThePikachu** [comment]
Also for authentication theres not much better you can do really. Maybe you can just do a bearer instead or go the ss14.watchdog route with basic auth (username and pass)

MoMMI authenticates the same way to the server too

File: Content.Server/Administration/ServerAPI.cs [hunk 1]
```diff
@@ -0,0 +1,694 @@
...
+            return false;
+        }
+
+        if (!CheckAccess(context))
+        {
+            await context.RespondErrorAsync(HttpStatusCode.Unauthorized);
+            return true;
+        }
+
+        var playerSupplied = context.RequestHeaders.TryGetValue("Username", out var username);
```
[the remark this answers is not in the document]
Comment (VasilisThePikachu) [reply Content.Server/Administration/ServerAPI.cs:379]: I actually suggested them to make this, we made a little chit chat in vc with pjb

Tldr my idea is to use get the list of active admins and players on the server and show their username in a hidden admin channel with the redbot discord bot. Now yes I can just query the auth API for this l. But now imagine doing that auth API for every single player just to resolve the guid to a player name... That will suck

File: Content.Server/Administration/ServerAPI.cs [hunk 2]
```diff
@@ -0,0 +1,694 @@
...
+        {
+            _sawmill.Info(@"Unauthorized access attempt to admin API. No auth header");
+            return false;
+        } // No auth header, no access
+
+        if (authToken == _token)
+            return true;
+
+        // Invalid auth header, no access
+        _sawmill.Info(@"Unauthorized access attempt to admin API. ""{0}"" vs ""{1}""", authToken, _token);
```
[the remark this answers is not in the document]
Comment (Simyon264) [reply Content.Server/Administration/ServerAPI.cs:670]: Watchdog does this as well. But I agree, logging the Token is bad.
```
_sawmill.Warning(
                    "received POST /shutdown with invalid authentication token. Ignoring {0}, {1}", auth,
                    _watchdogToken);
```
